# Assignment No. 2 - Emujin Batzorig



In [34]:
# import libraries
import pandas as pd
import os
os.getcwd()
from ucimlrepo import fetch_ucirepo 


## First Dataset - Adult Dataset from UCI

In [35]:
adult_data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data")
print(adult_data.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32560 entries, 0 to 32559
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   39              32560 non-null  int64 
 1    State-gov      32560 non-null  object
 2    77516          32560 non-null  int64 
 3    Bachelors      32560 non-null  object
 4    13             32560 non-null  int64 
 5    Never-married  32560 non-null  object
 6    Adm-clerical   32560 non-null  object
 7    Not-in-family  32560 non-null  object
 8    White          32560 non-null  object
 9    Male           32560 non-null  object
 10   2174           32560 non-null  int64 
 11   0              32560 non-null  int64 
 12   40             32560 non-null  int64 
 13   United-States  32560 non-null  object
 14   <=50K          32560 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None


In [36]:
adult_data.head()

,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
0,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
1,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
2,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
3,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
4,37,Private,284582,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K


## Question 1

### Question: Among the people with master's degree, are there a difference in the income between genders?

### Pseudocode: 

- Isolate people with master's degree
- Group by gender
- Group by gender and count the number of people in total
- Group by gender and count the number of people who earns >50K
- Find the proportions by dividing the number of people who earns more than 50k by the total count


Preparing the data frame first: 

In [37]:
# Renaming columns
adult_data.rename(columns={' Male': 'Gender', ' <=50K': 'Income Level', ' Never-married': 'Marriage Status', ' Bachelors': 'Education Level', ' Adm-clerical': 'Occupation'}, inplace=True)
adult_data.head(3)

,39,State-gov,77516,Education Level,13,Marriage Status,Occupation,Not-in-family,White,Gender,2174,0,40,United-States,Income Level
0,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
1,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
2,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K


In [38]:
# Changing data types of relevant columns
adult_data = adult_data.astype({'Gender': 'category', 'Income Level': str, 'Marriage Status': 'category', 'Education Level': str})

print(adult_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32560 entries, 0 to 32559
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   39               32560 non-null  int64   
 1    State-gov       32560 non-null  object  
 2    77516           32560 non-null  int64   
 3   Education Level  32560 non-null  object  
 4    13              32560 non-null  int64   
 5   Marriage Status  32560 non-null  category
 6   Occupation       32560 non-null  object  
 7    Not-in-family   32560 non-null  object  
 8    White           32560 non-null  object  
 9   Gender           32560 non-null  category
 10   2174            32560 non-null  int64   
 11   0               32560 non-null  int64   
 12   40              32560 non-null  int64   
 13   United-States   32560 non-null  object  
 14  Income Level     32560 non-null  object  
dtypes: category(2), int64(6), object(7)
memory usage: 3.3+ MB
None


### Single Line

Creating a new data frame with only people with master's degree

In [39]:
adult_masters = adult_data.query("`Education Level` == ' Masters'")
print(len(adult_masters))
adult_masters.head(5)


1723


,39,State-gov,77516,Education Level,13,Marriage Status,Occupation,Not-in-family,White,Gender,2174,0,40,United-States,Income Level
4,37,Private,284582,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K
7,31,Private,45781,Masters,14,Never-married,Prof-specialty,Not-in-family,White,Female,14084,0,50,United-States,>50K
18,43,Self-emp-not-inc,292175,Masters,14,Divorced,Exec-managerial,Unmarried,White,Female,0,0,45,United-States,>50K
46,44,Private,128354,Masters,14,Divorced,Exec-managerial,Unmarried,White,Female,0,0,40,United-States,<=50K
86,33,Private,202051,Masters,14,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K


In [40]:
# Counting how many people are in each gender regardless of income
# .size() counts the number of rows when called on a GroupBy object 
total_count_bygender = adult_masters.groupby('Gender').size()

/tmp/ipykernel_682/3703371199.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  total_count_bygender = adult_masters.groupby('Gender').size()


In [41]:
# Counting the number of people with income above 50k in each gender

high_income_count = adult_masters[adult_masters['Income Level'] == ' >50K'].groupby('Gender').size()

/tmp/ipykernel_682/1986305440.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  high_income_count = adult_masters[adult_masters['Income Level'] == ' >50K'].groupby('Gender').size()


In [42]:
proportion = high_income_count / total_count_bygender
print(proportion)


Gender
Female    0.333955
Male      0.657119
dtype: float64


### Function:

In [43]:
def income_gap(x):
    # x1 is the masters_data where we isolated people with masters degree
    x1 = x.query("`Education Level` == ' Masters'")
    total1 = x1.groupby('Gender').size() # counting by gender
    total2 = x1[x1['Income Level'] == ' >50K'].groupby('Gender').size() # counting by gender and high income
    p = (total2 / total1)
    
    return p


    

In [44]:
print(income_gap(adult_data))

Gender
Female    0.333955
Male      0.657119
dtype: float64


/tmp/ipykernel_682/34581761.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  total1 = x1.groupby('Gender').size() # counting by gender
/tmp/ipykernel_682/34581761.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  total2 = x1[x1['Income Level'] == ' >50K'].groupby('Gender').size() # counting by gender and high income


Answer to Q1: The proportion suggests that about 65.7% of Males make more than $50K a year while only 33.3% of Females make more than $50K a year even though they have the same education level (masters degree). Among the people with master's degree, there is a significant difference in the income between genders.

## Question 2

### Question: Are there differences in the number of people who make over $50k by Marriage Status?

target variable - ratio of people who make over $50K

### Pseudocode: 

- Group by marriage status
- Count total high income people
- Calculate % high income people in each groups

### Single Line: 

In [45]:
# Slicing the main dataframe to make it simpler
adult1 = adult_data[['Marriage Status', 'Income Level']]

In [46]:
adult2 = adult1.groupby('Marriage Status')

# Initializing a new DataFrame
df = pd.DataFrame()
df = df.assign(marriagestat_total=adult2.size(), # Total number of people in each groups
               
               # Number of people who earn >50K in each group
            income_count = (adult1['Income Level'] == ' >50K').groupby(adult1['Marriage Status']).sum(),
            
               # Calculating the % in each group
            percentage = lambda df: (df['income_count'] / df['marriagestat_total']))
df = df.sort_values(by='percentage')
print(df)

                       marriagestat_total  income_count  percentage
Marriage Status                                                    
Never-married                       10682           491    0.045965
Separated                            1025            66    0.064390
Married-spouse-absent                 418            34    0.081340
Widowed                               993            85    0.085599
Divorced                             4443           463    0.104209
Married-AF-spouse                      23            10    0.434783
Married-civ-spouse                  14976          6692    0.446848


/tmp/ipykernel_682/1257882556.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  adult2 = adult1.groupby('Marriage Status')
/tmp/ipykernel_682/1257882556.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  income_count = (adult1['Income Level'] == ' >50K').groupby(adult1['Marriage Status']).sum(),


### Function: 

In [47]:
def income_gap(data):
    adult1 = data[['Marriage Status', 'Income Level']]
    adult2 = adult1.groupby('Marriage Status')
    df = pd.DataFrame()
    df = df.assign(marriagestat_total=adult2.size(),
                   income_count = (adult1['Income Level'] == ' >50K').groupby(adult1['Marriage Status']).sum(),
                   percentage = lambda df: (df['income_count'] / df['marriagestat_total']))
    df = df.sort_values(by='percentage')
    
    return df

In [48]:
income_gap(adult_data)

/tmp/ipykernel_682/4023702366.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  adult2 = adult1.groupby('Marriage Status')
/tmp/ipykernel_682/4023702366.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  income_count = (adult1['Income Level'] == ' >50K').groupby(adult1['Marriage Status']).sum(),


,marriagestat_total,income_count,percentage
Marriage Status,,,
Never-married,10682,491,0.045965
Separated,1025,66,0.064390
Married-spouse-absent,418,34,0.081340
Widowed,993,85,0.085599
Divorced,4443,463,0.104209
Married-AF-spouse,23,10,0.434783
Married-civ-spouse,14976,6692,0.446848


Answer to Q2: The dataframe suggests that on average, people who are married to a member of the Armed Forces or a Civilian are likely to earn >50K than people who have different marriage statuses. According to the data, people who were never-married are the least likely to earn over 50K. 

## Second Dataset - Github Stats

About the dataset: 
- name: programming language
- year: year the data was recorded
- quarter: quarter of the year
- count: the number of pull requests made 

In [162]:
github_stats = pd.read_csv('https://raw.githubusercontent.com/MainakRepositor/Datasets/refs/heads/master/Github_Stats/prs.csv')
print(github_stats.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3462 entries, 0 to 3461
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     3462 non-null   object
 1   year     3462 non-null   int64 
 2   quarter  3462 non-null   int64 
 3   count    3462 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 108.3+ KB
None


## Question 3
### Question: Do pull requests vary by quarters within a year? 

### Sudo code: 
- Change data type to categorical
- Group by quarter
- Create a new dataframe
- Compute the sum of pull requests in each group

### Single Line: 

In [163]:
# Convert the data type first so we can group 
github_stats['quarter'] = github_stats['quarter'].astype("category")

github_df = pd.DataFrame()

# Group by the quarters and compute sum of the count column
github_df = github_df.assign(
    total_pull = github_stats.groupby('quarter')['count'].sum()
)

github_df = github_df.sort_values(by='total_pull')

print(github_df)

         total_pull
quarter            
3           9531436
1           9608930
2           9687025
4           9710696


/tmp/ipykernel_682/2044632903.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  total_pull = github_stats.groupby('quarter')['count'].sum()


### Function

In [165]:
def func4(data3): 
    data3['quarter'] = data3['quarter'].astype("category")
    github_df = pd.DataFrame()
    github_df = github_df.assign(total_pull = data3.groupby('quarter')['count'].sum())
    github_df = github_df.sort_values(by='total_pull')
    
    return github_df

print(func4(github_stats))





         total_pull
quarter            
3           9531436
1           9608930
2           9687025
4           9710696


/tmp/ipykernel_682/3062322404.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  github_df = github_df.assign(total_pull = data3.groupby('quarter')['count'].sum())


Answer to Q3: The data tells us that, the sum of the pull requests made in each quarters do not have a significant difference. 

## Question 4
### Question: Do the number of pull requests made in Github vary by programming languages? 

### Sudo code: 
- Group by programming language names
- Compute the sum of pull requests made in each groups
- Add the result to a new dataframe

### Single Line: 

In [174]:
# Changing the data type
github_stats['name'] = github_stats['name'].astype("category")

# Initialize a new data frame
github_df2 = pd.DataFrame()

# Group by the 'name' and compute sum of the 'count' column & assign it to the new df
github_df2 = github_df2.assign(
    pull_lang = github_stats.groupby('name')['count'].sum()
)

github_d2 = github_df2.sort_values(by='pull_lang')

print(github_df2)



               pull_lang
name                    
1C Enterprise        329
ABAP                4743
AGS Script           139
API Blueprint       1066
ASP                 1048
...                  ...
Zephir               186
mIRC Script         7877
nesC                4459
ooc                 4120
q                    360

[187 rows x 1 columns]


/tmp/ipykernel_682/695356200.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  pull_lang = github_stats.groupby('name')['count'].sum()


### Function: 

In [172]:
def func5(data4): 
    data4['name'] = data4['name'].astype("category")
    github_df2 = pd.DataFrame()
    github_df2 = github_df2.assign(pull_lang = data4.groupby('name')['count'].sum())
    github_df2 = github_df2.sort_values(by='pull_lang', ascending=False)
    
    return github_df2

print(func5(github_stats))

            pull_lang
name                 
JavaScript    7512246
Python        5904273
Java          3731712
Ruby          2887567
PHP           2474738
...               ...
Ceylon            111
CWeb              107
Blade             107
Modelica          107
MoonScript        101

[187 rows x 1 columns]


/tmp/ipykernel_682/2818401044.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  github_df2 = github_df2.assign(pull_lang = data4.groupby('name')['count'].sum())


Answer to Question 4: From the output of the function, we can see that there are 187 rows, each row representing a programming language. The most commonly used programming languages used in pull request are JavaScript, Python, and Java. Additionally, we can see that Modelica and MoonScript have the least number of pull request. 